# Building a Spam Filter with Niave Bayes

In this project we will classify messages as spam or non spam using a native Bayes Filter

The dataset can be downloaded from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection) and the data collection process is described [here](http://www.dt.fee.unicamp.br/~tiago/smsspamcollection/#composition)

In [1]:
import pandas as pd
sms_spam = pd.read_csv('SMSSpamCollection',sep='\t',header=None,names=['Label','SMS'])

In [2]:
print(sms_spam.shape)
sms_spam.head()

(5572, 2)


Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [3]:
sms_spam['Label'].value_counts(normalize=True)

ham     0.865937
spam    0.134063
Name: Label, dtype: float64

Roughly 87% of messages are ham with with 13% of messages as spam.  This is in line with what we would expect as most messages received are non spam (ham)

# Training and Test Set

We will split the data into two sets a training set with 80% (4458 messages) and a test set with 20% (1,114) of the dataset

In [4]:
# Random dataset
data_random = sms_spam.sample(frac=1,random_state=1)

In [5]:
data_random.head()

Label                                                SMS
1078   ham                       Yep, by the pretty sculpture
4028   ham      Yes, princess. Are you going to make me moan?
958    ham                         Welp apparently he retired
4642   ham                                            Havent.
4674   ham  I forgot 2 ask ü all smth.. There's a card on ...

In [6]:
# Get the index for the training set it is 0.8 because that is 80% of the sample
training_index = round(len(data_random) * 0.8)
print(training_index)

4458


In [7]:
# Total amount of messages in sample
print(len(data_random))

5572


In [8]:
data_random[:training_index] # training set slice need to remove index as column

Label                                                SMS
1078   ham                       Yep, by the pretty sculpture
4028   ham      Yes, princess. Are you going to make me moan?
958    ham                         Welp apparently he retired
4642   ham                                            Havent.
4674   ham  I forgot 2 ask ü all smth.. There's a card on ...
5461   ham  Ok i thk i got it. Then u wan me 2 come now or...
4210   ham  I want kfc its Tuesday. Only buy 2 meals ONLY ...
4216   ham                         No dear i was sleeping :-P
1603   ham                          Ok pa. Nothing problem:-)
1504   ham                    Ill be there on  &lt;#&gt;  ok.
1783   ham  My uncles in Atlanta. Wish you guys a great se...
3465   ham                                           My phone
5534   ham                       Ok which your another number
4267   ham  The greatest test of courage on earth is to be...
2498   ham  Dai what this da.. Can i send my resume to thi...
4259   ham                      I am late. I will be there at
147   spam  FreeMsg Why haven't you replied to my text? I'...
141    ham                  K, text me when you're on the way
4517  spam  Congrats! 2 mobile 3G Videophones R yours. cal...
3053   ham  Please leave this topic..sorry for telling that..
5392   ham  Ooooooh I forgot to tell u I can get on yovill...
2346   ham          Hi this is yijue, can i meet u at 11 tmr?
1242   ham  I want to show you the world, princess :) how ...
3224   ham                  Well that must be a pain to catch
4872   ham                Well. You know what i mean. Texting
3044   ham         Your bill at 3 is £33.65 so thats not bad!
1660   ham                       Yeah, where's your class at?
3214   ham                                     What's ur pin?
501    ham  Fighting with the world is easy, u either win ...
1827   ham  Dude. What's up. How Teresa. Hope you have bee...
...    ...                                                ...
3117   ham                Uncle Abbey! Happy New Year. Abiola
2020   ham             From tomorrow onwards eve 6 to 3 work.
4827   ham                      Haha, just what I was thinkin
2974   ham                           Happy New Year Princess!
1275   ham  Let me know how to contact you. I've you settl...
2413  spam  I don't know u and u don't know me. Send CHAT ...
528    ham         Yes! How is a pretty lady like you single?
4404   ham                             Just getting back home
1505   ham                         Oh my God. I'm almost home
3598  spam  Congratulations YOU'VE Won. You're a Winner in...
3681   ham  I cant pick the phone right now. Pls send a me...
1318  spam  Win the newest “Harry Potter and the Order of ...
3380   ham                    Dear umma she called me now :-)
3034   ham                        Aight, lemme know what's up
2380   ham  Good evening Sir, hope you are having a nice d...
4311  spam  Someone U know has asked our dating service 2 ...
2616   ham         2marrow only. Wed at  &lt;#&gt;  to 2 aha.
1166   ham        Haha yeah I see that now, be there in a sec
3471   ham                        aathi..where are you dear..
4695   ham  Pls give her the food preferably pap very slow...
5293   ham             I donno its in your genes or something
2871  spam  YOUR CHANCE TO BE ON A REALITY FANTASY SHOW ca...
1222   ham                             Prakesh is there know.
4749   ham  The beauty of life is in next second.. which h...
4255   ham             How about clothes, jewelry, and trips?
1982   ham  Sorry, I'll call later in meeting any thing re...
5180   ham  Babe! I fucking love you too !! You know? Fuck...
4020  spam  U've been selected to stay in 1 of 250 top Bri...
371    ham  Hello my boytoy ... Geeee I miss you already a...
3482   ham                           Wherre's my boytoy ? :-(

[4458 rows x 2 columns]

In [9]:
# create both training and test sets using slices
training_set = data_random[:training_index].reset_index(drop=True)
test_set = data_random[training_index:].reset_index(drop=True)
print(training_set.shape)
print(test_set.shape)

(4458, 2)
(1114, 2)


From the previous section we found that roughly 87% of messages are ham and 13% are spam let's see if that holds up in the training and test data sets

In [10]:
training_set['Label'].value_counts(normalize=True)

ham     0.86541
spam    0.13459
Name: Label, dtype: float64

In [11]:
test_set['Label'].value_counts(normalize=True)

ham     0.868043
spam    0.131957
Name: Label, dtype: float64

The results validate our current results

# Letter Case and Punctuation

We will need to clean the data before we can proceed.  We want to bring the table to a version that we can count words from a vocabulary.  We also are going to remove all punctuation and make all words lowercase

In [12]:
# Current form (using training set)
training_set.head()

Label                                                SMS
0   ham                       Yep, by the pretty sculpture
1   ham      Yes, princess. Are you going to make me moan?
2   ham                         Welp apparently he retired
3   ham                                            Havent.
4   ham  I forgot 2 ask ü all smth.. There's a card on ...

In [13]:
training_set['SMS'] = training_set['SMS'].str.replace('\W',' ')
training_set['SMS'] = training_set['SMS'].str.lower()
training_set.head()

Label                                                SMS
0   ham                       yep  by the pretty sculpture
1   ham      yes  princess  are you going to make me moan 
2   ham                         welp apparently he retired
3   ham                                            havent 
4   ham  i forgot 2 ask ü all smth   there s a card on ...

# Creating the Vocabulary

In [14]:
training_set['SMS'] = training_set['SMS'].str.split()

In [15]:
vocabulary = []
for sms in training_set['SMS']:
    for word in sms:
        vocabulary.append(word)

vocabulary = list(set(vocabulary))

In [16]:
len(vocabulary)

7783

Our vocabulary has 7,783 unique words in it

# The Final Training Set

Now that we have the vocabulary list we can complete the transformation of the training dataset

In [17]:
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1

In [18]:
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()

0  00  000  000pes  008704050406  0089  01223585334  02  0207  02072069400  \
0  0   0    0       0             0     0            0   0     0            0   
1  0   0    0       0             0     0            0   0     0            0   
2  0   0    0       0             0     0            0   0     0            0   
3  0   0    0       0             0     0            0   0     0            0   
4  0   0    0       0             0     0            0   0     0            0   

  ...  zindgi  zoe  zogtorius  zouk  zyada  é  ú1  ü  〨ud  鈥  
0 ...       0    0          0     0      0  0   0  0    0  0  
1 ...       0    0          0     0      0  0   0  0    0  0  
2 ...       0    0          0     0      0  0   0  0    0  0  
3 ...       0    0          0     0      0  0   0  0    0  0  
4 ...       0    0          0     0      0  0   0  2    0  0  

[5 rows x 7783 columns]

In [19]:
training_set_clean = pd.concat([training_set,word_counts],axis=1)
training_set_clean.head()

Label                                                SMS  0  00  000  \
0   ham                  [yep, by, the, pretty, sculpture]  0   0    0   
1   ham  [yes, princess, are, you, going, to, make, me,...  0   0    0   
2   ham                    [welp, apparently, he, retired]  0   0    0   
3   ham                                           [havent]  0   0    0   
4   ham  [i, forgot, 2, ask, ü, all, smth, there, s, a,...  0   0    0   

   000pes  008704050406  0089  01223585334  02 ...  zindgi  zoe  zogtorius  \
0       0             0     0            0   0 ...       0    0          0   
1       0             0     0            0   0 ...       0    0          0   
2       0             0     0            0   0 ...       0    0          0   
3       0             0     0            0   0 ...       0    0          0   
4       0             0     0            0   0 ...       0    0          0   

   zouk  zyada  é  ú1  ü  〨ud  鈥  
0     0      0  0   0  0    0  0  
1     0      0  0   0  0    0  0  
2     0      0  0   0  0    0  0  
3     0      0  0   0  0    0  0  
4     0      0  0   0  2    0  0  

[5 rows x 7785 columns]

# Calculating Constants First
Here we will calculate P(Spam) and P(Ham) along with N_Spam, N_Ham, and N_Vocabulary

In [22]:
spam_messages = training_set_clean[training_set_clean['Label'] == 'spam']
ham_messages = training_set_clean[training_set_clean['Label'] == 'ham']

# Laplace smoothing
alpha = 1

# P(Spam) and P(Ham)
p_spam = len(spam_messages) / len(training_set_clean) # Number of spam messages over total messages
p_ham = len(ham_messages) / len(training_set_clean)

# N Spam (words)
n_words_per_spam_message = spam_messages['SMS'].apply(len)
n_spam = n_words_per_spam_message.sum()

# N Ham
n_words_per_ham_message = ham_messages['SMS'].apply(len)
n_ham = n_words_per_ham_message.sum()

# N Vocabulary length of vocabulary
n_vocabulary = len(vocabulary)

# Calculating Parameters
Here we will calculate P(w_i|Spam) and P(w_i|Ham) using the constants defineded in the previous cell

In [23]:
parameters_spam = {unique_word:0 for unique_word in vocabulary}
parameters_ham = {unique_word:0 for unique_word in vocabulary}

In [24]:
for word in vocabulary:
    n_word_given_spam = spam_messages[word].sum()
    p_word_given_spam = (n_word_given_spam + alpha) / (n_spam + alpha * n_vocabulary)
    parameters_spam[word] = p_word_given_spam
    
    n_word_given_ham = ham_messages[word].sum()
    p_word_given_ham = (n_word_given_ham + alpha) / (n_ham + alpha * n_vocabulary)
    parameters_ham[word] = p_word_given_ham

# Classifying A New Message

In [30]:
import re

def classify(message):

    message = re.sub('\W', ' ', message)
    message = message.lower()
    message = message.split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham
    
    for word in message:
        if word in parameters_spam:
            p_spam_given_message *=parameters_spam[word]
        
        if word in parameters_ham:
            p_ham_given_message *=parameters_ham[word]
   

    print('P(Spam|message):', p_spam_given_message)
    print('P(Ham|message):', p_ham_given_message)

    if p_ham_given_message > p_spam_given_message:
        print('Label: Ham')
    elif p_ham_given_message < p_spam_given_message:
        print('Label: Spam')
    else:
        print('Equal proabilities, have a human classify this!')

In [31]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')

P(Spam|message): 1.3481290211300841e-25
P(Ham|message): 1.9368049028589875e-27
Label: Spam


In [32]:
classify("Sounds good, Tom, then see u there")

P(Spam|message): 2.4372375665888117e-25
P(Ham|message): 3.687530435009238e-21
Label: Ham


# Measuring the Spam Filter's Accuracy
We will now be using the testing dataset to verify the accuracy of the spam filter

In [34]:
def classify_test_set(message):

    message = re.sub('\W', ' ', message)
    message = message.lower()
    message = message.split()

    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]

        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]

    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_spam_given_message > p_ham_given_message:
        return 'spam'
    else:
        return 'needs human classification'

Using this new function we will add a new column called predicted to the test dataset

In [35]:
test_set['predicted'] = test_set['SMS'].apply(classify_test_set)
test_set.head()

Label                                                SMS predicted
0   ham          Later i guess. I needa do mcat study too.       ham
1   ham             But i haf enuff space got like 4 mb...       ham
2  spam  Had your mobile 10 mths? Update to latest Oran...      spam
3   ham  All sounds good. Fingers . Makes it difficult ...       ham
4   ham  All done, all handed in. Don't know if mega sh...       ham

In [37]:
correct = 0
total = test_set.shape[0] # rows of test set only

# Iterating over all rows
for row in test_set.iterrows():
    row = row[1]
    if row['Label'] == row['predicted']:
        correct +=1
        
incorrect = total - correct
accuracy = correct / total
print('Correct:',correct)
print('Incorrect:',incorrect)
print('Accuracy:',accuracy)

Correct: 1100
Incorrect: 14
Accuracy: 0.9874326750448833


The accuracy is really good with a 98.74% rate classifying 1100 messages as correct

# Next Steps
In this project, we managed to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm. The filter had an accuracy of 98.74% on the test set, which is an excellent result. We initially aimed for an accuracy of over 80%, but we managed to do way better than that.

Next steps are
* Isolating the 14 messages that are not correct and figure out why they are incorrect
* Make the filtering process more complex by making the algorithm sensitive to letter case